docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,153 @@
+
+## Desired layers
+
+### Transport
+
+The transport layer is concerned with round-tripping requests to an apiserver
+somewhere. It consumes a Config object with options appropriate for this.
+(That's most of the current client.Config structure.)
+
+Third parties (e.g., open shift) may have their own implementation.
+
+Transport delivers an object that implements http's RoundTripper interface
+and/or can be used in place of http.DefaultClient to route requests.
+
+Transport objects are safe for concurrent use, and are cached and reused by
+subsequent layers.
+
+Tentative name: "Transport".
+
+Action items:
+* Split out of current client package into a new package. (@krousey)
+
+### RESTClient/request.go
+
+RESTClient consumes a Transport, a group, and a version, and produces something
+that implements the interface currently in request.go. That is, with a
+RESTClient, you can write chains of calls like: c.Get().Path(p).Param("name",
+"value").Do()
+
+The group and version may be empty when constructing a RESTClient. This is valid
+for executing discovery commands. The group and version may be overridable with
+a function call.
+
+RESTClients may have semantic information built in. For example, if the
+query parameter changes from "labels" to "label-selector", when you call the
+RESTClient construction logic, it should give you a RESTClient which uses the
+appropriate query parameter when you call the "LabelSelector()" method.
+
+3rd parties such as open shift may provide their own implementation, but it
+should conform to the same interface as the standard RESTClient.
+
+Action items:
+* Split out of the current client package. (@krousey)
+* Convert to an interface (currently, it's a struct).
+
+### Mux layer
+
+The client muxer layer has a map of group/version to cached RESTClient, and
+knows how to construct a new RESTClient in case of a cache miss.
+
+Tentative name: MetaClient
+
+Action items:
+* Move client cache out of kubectl libraries into a more general home.
+
+### High-level: Individual typed
+
+Our current high-level client allows you to write things like
+`c.Pods("namespace").Create(p)`; we will insert a level for the group.
+
+That is, the system will be:
+`client.GroupName().NamespaceSpecifier().RESTAction()`
+
+The types returned/consumed by such functions will be e.g. api/v1, NOT the
+current version inspecific types.
+
+We will provide this structure for each version of each group. It is infeasible
+to do this manually, so we will generate this. The generator will accept both
+swagger and the ordinary go types.
+
+Typed clients will be constructabale given a MetaClient.
+
+Action items:
+* Move current typed clients into new directory structure (described below)
+* Finish client generation logic. (@lavalamp)
+
+#### High-level, typed: Discovery
+
+A `DiscoveryClient` is necessary to discover the api groups, versions, and
+resources a server supports. It's constructable given a RESTClient. It is
```

Comment:
RESTClient does not require apiVersion.  The absence of an apiversion is
not an error. It should use codec - why does discovery have to be tied to
JSON?  And codec being absent could very well support the null codec (or a
simple JSON codec)

RESTClient exists to abstract access to RESTful apis.  It also exists to
unify client-server interactions over HTTP with conventions.  Having two
code paths to communicate to a server is not desirable.

On Mon, Oct 19, 2015 at 4:52 PM, Chao Xu <notifications@github.com> wrote:

> In docs/proposals/client-package-structure.md
> <https://github.com/kubernetes/kubernetes/pull/15730#discussion_r42424233>
> :
>
> > +current version inspecific types.
> > +
> > +We will provide this structure for each version of each group. It is infeasible
> > +to do this manually, so we will generate this. The generator will accept both
> > +swagger and the ordinary go types.
> > +
> > +Typed clients will be constructabale given a MetaClient.
> > +
> > +Action items:
> > +* Move current typed clients into new directory structure (described below)
> > +* Finish client generation logic. (@lavalamp)
> > +
> > +#### High-level, typed: Discovery
> > +
> > +A `DiscoveryClient` is necessary to discover the api groups, versions, and
> > +resources a server supports. It's constructable given a RESTClient. It is
>
> RESTClient has apiVersion and codec. The discovery client shouldn't have
> those, it's version agnostic and the messages it handles are not encoded to
> a version.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15730/files#r42424233>.
>